// Copyright 2015, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";

package csce438;

import "google/protobuf/timestamp.proto";

// The messenger service definition.
service SNSService{

  rpc Login (Request) returns (Reply) {}
  rpc List (Request) returns (ListReply) {}
  rpc Follow (Request) returns (Reply) {}
  rpc UnFollow (Request) returns (Reply) {}
  // Bidirectional streaming RPC
  rpc Timeline (stream Message) returns (stream Message) {} 
  rpc changeServerType(stream Request) returns (stream Request){} // it doesn't matter
  rpc getForeignClients (Request) returns (Reply) {}
  rpc getListOfClients (Request) returns (Reply) {} // this is between the synchronizers


  // these ones below are for the synchronizer
  rpc getServerClients (Request) returns (Reply) {} // the synchronizer needs a list of 
  //all the clients that are present in a cluster

  rpc sendListOfClients (Request) returns (Reply) {} // synchronizer sends list of foreign clients to the master
  // sendListOfClients will also have a reply with requests that need to be made of other servers for updates

  rpc sendServerForeignClients (Request) returns (Reply) {} // synchronizer needs to update 
  // the servers about which clients can be followed from other clusters

  rpc sendClientUpdates (Request) returns (Reply) {}
  // this will be called after sendListofClients returns requests For
  //updates from foreign clients, and will reply with those updates accordingly 

  //what data type I use here, if this happens then I know to change the type of the server

  rpc sendFollows (Request) returns (Reply) {} // let the master be able to know which clients want to follow
  // out of cluster clients
}

message ListReply {
  repeated string all_users = 1;
  repeated string followers = 2;
}

message Request {
  string username = 1;
  repeated string arguments = 2;
}

message Reply {
  string msg = 1;
}

message Message {
  //Username who sent the message
  string username = 1;
  //Message from the user
  string msg = 2;
  //Time the message was sent
  google.protobuf.Timestamp timestamp = 3;
}
